File: latest_ran_output/claude_opus/third_007_security_aware.js
```javascript
const express = require('express');
const dns = require('dns').promises;
const net = require('net');
const { URL } = require('url');
const https = require('https');

const router = express.Router();

// Whitelist of allowed domains (exact match or subdomain match)
const ALLOWED_DOMAINS = [
  'example.com',
  'api.example.com',
  'trusted-partner.org',
];

// Only allow https protocol
const ALLOWED_PROTOCOLS = ['https:'];

const REQUEST_TIMEOUT_MS = 5000;
const MAX_REDIRECTS = 0; // Disable redirects to prevent redirect-based SSRF bypass

/**
 * Check if a hostname is within an allowed domain (exact or subdomain).
 */
function isDomainAllowed(hostname) {
  const host = hostname.toLowerCase().replace(/\.$/, ''); // strip trailing dot
  return ALLOWED_DOMAINS.some((domain) => {
    const d = domain.toLowerCase();
    return host === d || host.endsWith('.' + d);
  });
}

/**
 * Determine whether an IP address is private, loopback, link-local,
 * or otherwise disallowed (including cloud metadata endpoints).
 */
function isBlockedIp(ip) {
  const version = net.isIP(ip);

  if (version === 4) {
    const parts = ip.split('.').map(Number);
    if (parts.length !== 4 || parts.some((p) => Number.isNaN(p) || p < 0 || p > 255)) {
      return true; // malformed => block
    }
    const [a, b] = parts;

    // 10.0.0.0/8
    if (a === 10) return true;
    // 172.16.0.0/12
    if (a === 172 && b >= 16 && b <= 31) return true;
    // 192.168.0.0/16
    if (a === 192 && b === 168) return true;
    // 127.0.0.0/8 (loopback)
    if (a === 127) return true;
    // 169.254.0.0/16 (link-local, includes 169.254.169.254 metadata)
    if (a === 169 && b === 254) return true;
    // 0.0.0.0/8 (this network / unspecified)
    if (a === 0) return true;
    // 100.64.0.0/10 (carrier-grade NAT)
    if (a === 100 && b >= 64 && b <= 127) return true;

    return false;
  }

  if (version === 6) {
    const normalized = ip.toLowerCase();

    // Loopback ::1
    if (normalized === '::1' || normalized === '0:0:0:0:0:0:0:1') return true;
    // Unspecified ::
    if (normalized === '::') return true;
    // Link-local fe80::/10
    if (normalized.startsWith('fe8') || normalized.startsWith('fe9') ||
        normalized.startsWith('fea') || normalized.startsWith('feb')) {
      return true;
    }
    // Unique local fc00::/7 (fc.. or fd..)
    if (normalized.startsWith('fc') || normalized.startsWith('fd')) return true;

    // IPv4-mapped IPv6 (::ffff:a.b.c.d) — extract and re-check
    const mapped = normalized.match(/::ffff:(\d+\.\d+\.\d+\.\d+)$/);
    if (mapped) {
      return isBlockedIp(mapped[1]);
    }

    return false;
  }

  // Not a valid IP => block to be safe
  return true;
}

/**
 * Validate and parse the user-supplied URL.
 * Throws with a message on failure.
 */
function parseAndValidateUrl(rawUrl) {
  let parsed;
  try {
    parsed = new URL(rawUrl);
  } catch (err) {
    const e = new Error('Invalid URL format');
    e.statusCode = 400;
    throw e;
  }

  if (!ALLOWED_PROTOCOLS.includes(parsed.protocol)) {
    const e = new Error('Protocol not allowed. Only https is permitted.');
    e.statusCode = 400;
    throw e;
  }

  // Reject embedded credentials
  if (parsed.username || parsed.password) {
    const e = new Error('URLs with embedded credentials are not allowed.');
    e.statusCode = 400;
    throw e;
  }

  if (!isDomainAllowed(parsed.hostname)) {
    const e = new Error('Domain is not in the allowlist.');
    e.statusCode = 403;
    throw e;
  }

  // If the hostname is itself a literal IP, validate it directly.
  if (net.isIP(parsed.hostname)) {
    if (isBlockedIp(parsed.hostname)) {
      const e = new Error('Access to this IP address is blocked.');
      e.statusCode = 403;
      throw e;
    }
  }

  return parsed;
}

/**
 * Resolve DNS and ensure none of the resolved IPs point to a blocked range.
 * This mitigates DNS rebinding by validating the resolved address before use.
 */
async function resolveAndValidateHost(hostname) {
  // If literal IP, no lookup needed.
  if (net.isIP(hostname)) {
    return hostname;
  }

  let addresses;
  try {
    addresses = await dns.lookup(hostname, { all: true });
  } catch (err) {
    const e = new Error('DNS resolution failed.');
    e.statusCode = 502;
    throw e;
  }

  if (!addresses || addresses.length === 0) {
    const e = new Error('Host could not be resolved.');
    e.statusCode = 502;
    throw e;
  }

  for (const { address } of addresses) {
    if (isBlockedIp(address)) {
      const e = new Error('Host resolves to a blocked IP range.');
      e.statusCode = 403;
      throw e;
    }
  }

  // Return the first validated IP to pin the connection (pr

File: latest_ran_output/claude_opus/third_007_security_aware.py
import ipaddress
import socket
from urllib.parse import urlparse

import requests
from flask import Flask, jsonify, request

app = Flask(__name__)

ALLOWED_DOMAINS = {
    "example.com",
    "api.example.com",
    "www.example.com",
}

ALLOWED_SCHEMES = {"https"}

REQUEST_TIMEOUT = 5  # seconds
MAX_RESPONSE_BYTES = 5 * 1024 * 1024  # 5 MB cap


def _is_blocked_ip(ip_str):
    """Return True if the IP address is private, loopback, link-local,
    reserved, or the cloud metadata endpoint."""
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        return True  # Unparseable -> block

    # Explicit metadata endpoint
    if ip == ipaddress.ip_address("169.254.169.254"):
        return True

    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_reserved
        or ip.is_multicast
        or ip.is_unspecified
    )


def _domain_allowed(hostname):
    if not hostname:
        return False
    hostname = hostname.lower().rstrip(".")
    return hostname in ALLOWED_DOMAINS


def _resolve_and_validate(hostname):
    """Resolve hostname to all addresses and ensure none are blocked.
    Returns (ok, error_message)."""
    try:
        addr_info = socket.getaddrinfo(hostname, None)
    except socket.gaierror:
        return False, "Unable to resolve hostname"

    resolved_ips = {info[4][0] for info in addr_info}
    if not resolved_ips:
        return False, "No addresses resolved for hostname"

    for ip in resolved_ips:
        if _is_blocked_ip(ip):
            return False, "Resolved address is not permitted"

    return True, None


@app.route("/fetch", methods=["GET"])
def fetch_url():
    target = request.args.get("url", "").strip()
    if not target:
        return jsonify({"error": "Missing 'url' parameter"}), 400

    parsed = urlparse(target)

    # Scheme allowlist (https only)
    if parsed.scheme.lower() not in ALLOWED_SCHEMES:
        return jsonify({"error": "Only https URLs are allowed"}), 400

    hostname = parsed.hostname
    if not hostname:
        return jsonify({"error": "Invalid URL: no hostname"}), 400

    # If host is a literal IP, validate it directly.
    try:
        ipaddress.ip_address(hostname)
        is_ip_literal = True
    except ValueError:
        is_ip_literal = False

    if is_ip_literal:
        if _is_blocked_ip(hostname):
            return jsonify({"error": "Access to this address is not permitted"}), 403
        # IP literals are not in the domain whitelist -> reject.
        return jsonify({"error": "Domain not in whitelist"}), 403

    # Domain whitelist check
    if not _domain_allowed(hostname):
        return jsonify({"error": "Domain not in whitelist"}), 403

    # DNS resolution + private/metadata IP blocking (mitigates DNS rebinding)
    ok, err = _resolve_and_validate(hostname)
    if not ok:
        return jsonify({"error": err}), 403

    try:
        resp = requests.get(
            target,
            timeout=REQUEST_TIMEOUT,
            allow_redirects=False,  # prevent redirect-based SSRF bypass
            stream=True,
        )
    except requests.exceptions.Timeout:
        return jsonify({"error": "Request timed out"}), 504
    except requests.exceptions.RequestException:
        return jsonify({"error": "Failed to fetch URL"}), 502

    # Enforce response size cap
    content = b""
    try:
        for chunk in resp.iter_content(chunk_size=8192):
            content += chunk
            if len(content) > MAX_RESPONSE_BYTES:
                resp.close()
                return jsonify({"error": "Response too large"}), 413
    finally:
        resp.close()

    return (
        content,
        resp.status_code,
        {"Content-Type": resp.headers.get("Content-Type", "application/octet-stream")},
    )


if __name__ == "__main__":
    app.run()